python/mxnet/ndarray.py

```diff
@@ -477,12 +480,35 @@ def reshape(self, shape):
         array([[ 0.,  1.],
                [ 2.,  3.],
                [ 4.,  5.]], dtype=float32)
+        >>> y = x.reshape((3,-1))
+        >>> y.asnumpy()
+        array([[ 0.,  1.],
+               [ 2.,  3.],
+               [ 4.,  5.]], dtype=float32)
         >>> y[:] = -1
         >>> x.asnumpy()
         array([[-1., -1., -1.],
                [-1., -1., -1.]], dtype=float32)
         """
         handle = NDArrayHandle()
+
+        # Infer the correct size for dim == -1
+        shape = list(shape)
+        for index, element in enumerate(shape):
+            if element == -1:
+                remainder = list(self.shape)
+                for i, e in enumerate(shape):  # pylint: disable=C0321
+                    if i != index and e == -1:
+                        raise ValueError('Only one dimension can be inferred.')
+                    try:
+                        remainder.remove(e)
```

Comment:
This wouldn't work as it would end up replacing `-1` by the product of the non `-1` dimensions in the reshape. We want to first remove the dimensions that are actually specified by the user before taking that product.
For instance, if `X.shape = (3, 4, 5, 2)`, then `X.reshape(4, 2, -1)` should translate to `X.reshape(4, 2, 3*5)`, not `X.reshape(4, 2, 4*2)`.